Duplicati/Library/Main/Operation/DeleteHandler.cs

```diff
@@ -92,19 +88,25 @@ public void DoRun(Database.LocalDeleteDatabase db, ref System.Data.IDbTransactio
                 var backend = bk ?? sharedManager;
 
                 if (!hasVerifiedBacked && !m_options.NoBackendverification)
-                    FilelistProcessor.VerifyRemoteList(backend, m_options, db, m_result.BackendWriter); 
-                
-                var filesetNumbers = db.FilesetTimes.Zip(Enumerable.Range(0, db.FilesetTimes.Count()), (a, b) => new Tuple<long, DateTime>(b, a.Value)).ToList();
-                var sets = db.FilesetTimes.Select(x => x.Value).ToArray();
-                var toDelete = GetFilesetsToDelete(db, sets);
+                    FilelistProcessor.VerifyRemoteList(backend, m_options, db, m_result.BackendWriter);
+
+                IListResultFileset[] filesets = db.FilesetsWithBackupVersion.ToArray();
+                List<IListResultFileset> versionsToDelete = new List<IListResultFileset>();
+                versionsToDelete.AddRange(new SpecificVersionsRemover(this.m_options).GetFilesetsToDelete(filesets));
+                versionsToDelete.AddRange(new KeepTimeRemover(this.m_options).GetFilesetsToDelete(filesets));
+                versionsToDelete.AddRange(new RetentionPolicyRemover(this.m_options).GetFilesetsToDelete(filesets));
 
-                if (!m_options.AllowFullRemoval && sets.Length == toDelete.Length)
+                // When determining the number of full versions to keep, we need to ignore the versions already marked for removal.
+                versionsToDelete.AddRange(new KeepVersionsRemover(this.m_options).GetFilesetsToDelete(filesets.Except(versionsToDelete)));
+
+                DateTime[] toDelete = versionsToDelete.Select(x => x.Time.ToUniversalTime()).Distinct().OrderByDescending(x => x).ToArray();

```

Comment:
Did you fix a bug here since the ordering wasn't there before? Anyway, if you only order to keep the latest one that you do within the if, then only do the sorting in the if. Can't you keep working with the variable versionsToDelete?

Reply:
This ordering was on line 255 of the old `GetFilesetsToDelete` method.  I wasn't sure if the `DropFilesetsFromTable` method assumed that the array was sorted.  I can look at that after I figure out why the test is failing.

Reply:
Revision a7ab5f6f00450dc39149addd028a24dbba66372f removes the unnecessary array creation and also removes some redundant `Timestamp` conversion (the conversion to universal time occurs in the `DropFilesetsFromTable` method.